docs/categories/02-Server/server-socket-instance.md

```diff
@@ -140,6 +140,39 @@ console.log(sockets[0].data.username); // "alice"
 
 More information [here](server-instance.md#utility-methods).
 
+:::caution
+Do not rely on reference equality when using properties of `data`.
+:::
+
+```js
+// BAD
+const activeUsers = new Set();
+io.on("connection", (socket) => {
+  socket.data.user = { name: "alice" };
+  activeUsers.add(socket.data.user);
+
+  socket.on("disconnect", () => {
+    // Socket.IO makes no guarantee data.user will reference the same object...
+    activeUsers.delete(socket.data.user);
+  });
+});
+```
+
+```js
+// GOOD
+const activeUsers = Map();
+io.on("connection", (socket) => {
+  const newUser = { name: "alice" };
+  socket.data.user = newUser;
+  activeUsers.set(socket.data.user.name, newUser);
+
+  socket.on("disconnect", () => {
+    // ... however primitive values will be preserved
+    activeUsers.delete(socket.data.user.name);
```

Comment:
Two things that bother me with this example:

- `socket.data.user.name` must be unique (implicitly)
- using a map here means it won't work when scaling to multiple servers

What do you think?

Reply:
Agreed, i've simply removed the 'Good' example. Hopefully just knowing it is bad will be enough to guide anyone who ran into the same problem as I did using a reference.